test/test_functional.py

```diff
@@ -183,5 +193,119 @@ def test_istft_of_sine(self):
         self._test_istft_of_sine(amplitude=99, L=10, n=7)
 
 
+def _num_stft_bins(signal_len, fft_len, hop_length, pad):
+    return (signal_len + 2 * pad - fft_len + hop_length) // hop_length
+
+
+@pytest.mark.parametrize('fft_length', [512])
+@pytest.mark.parametrize('hop_length', [256])
+@pytest.mark.parametrize('waveform', [
+    (torch.randn(1, 100000)),
+    (torch.randn(1, 2, 100000)),
+    pytest.param(torch.randn(1, 100), marks=xfail(raises=RuntimeError)),
+])
+@pytest.mark.parametrize('pad_mode', [
+    # 'constant',
+    'reflect',
+])
+@unittest.skipIf(not IMPORT_LIBROSA, 'Librosa is not available')
+def test_stft(waveform, fft_length, hop_length, pad_mode):
+    """
+    Test STFT for multi-channel signals.
+
+    Padding: Value in having padding outside of torch.stft?
+    """
+    pad = fft_length // 2
+    window = torch.hann_window(fft_length)
+    complex_spec = F.stft(waveform,
+                          fft_length=fft_length,
+                          hop_length=hop_length,
+                          window=window,
+                          pad_mode=pad_mode)
+    mag_spec, phase_spec = F.magphase(complex_spec)
+
+    # == Test shape
+    expected_size = list(waveform.size()[:-1])
+    expected_size += [fft_length // 2 + 1, _num_stft_bins(
+        waveform.size(-1), fft_length, hop_length, pad), 2]
+    assert complex_spec.dim() == waveform.dim() + 2
+    assert complex_spec.size() == torch.Size(expected_size)
+
+    # == Test values
+    fft_config = dict(n_fft=fft_length, hop_length=hop_length, pad_mode=pad_mode)
+    # note that librosa *automatically* pad with fft_length // 2.
+    expected_complex_spec = np.apply_along_axis(librosa.stft, -1,
+                                                waveform.numpy(), **fft_config)
+    expected_mag_spec, _ = librosa.magphase(expected_complex_spec)
+    # Convert torch to np.complex
+    complex_spec = complex_spec.numpy()
+    complex_spec = complex_spec[..., 0] + 1j * complex_spec[..., 1]
+
+    assert np.allclose(complex_spec, expected_complex_spec, atol=1e-5)
+    assert np.allclose(mag_spec.numpy(), expected_mag_spec, atol=1e-5)
+
+
+@pytest.mark.parametrize('rate', [0.5, 1.01, 1.3])
+@pytest.mark.parametrize('complex_specgrams', [
+    torch.randn(1, 2, 1025, 400, 2),
+    torch.randn(1, 1025, 400, 2)
+])
+@pytest.mark.parametrize('hop_length', [256])
+@unittest.skipIf(not IMPORT_LIBROSA, 'Librosa is not available')
+def test_phase_vocoder(complex_specgrams, rate, hop_length):
+
+    class use_double_precision:
```

Comment:
What's the advantage of this? Do we still plan to remove it?

Reply:
It's needed to compare the results of the `phase_vocoder` against `librosa`. Because it uses a cumulative sum, small numerical errors when not using double precision make them accumulate and differ. Some discussion on this [here](https://github.com/keunwoochoi/torchaudio-contrib/pull/65#discussion_r291855782).

Reply:
Can this only done using the "with" construct or can the dtype be passed explicitly?

Reply:
I think so, because `F.phase_vocoder` create a tensor [`time_steps`](https://github.com/pytorch/audio/blob/c8596a683f1ccdd53e8e1c31fa4c86974d81bba4/torchaudio/functional.py#L573) and that needs to be a `double()` if we want to compare it with `librosa`. Doing the `with` ensures that auxiliary tensors created by the phase vocoder have the correct precision.

Reply:
If that is the fear, then I think we should change phase_vocoder overall to respect the type of the input. Because, for testing this might be fine, but what if a user wants to do this herself and isn't aware of the need to create this context? Wanting to use doubles is still a legitimate feature and also being invariant under datatypes makes it easier to later on add float16 etc.

Reply:
Ok I'll make it so they have the same dtype as the input tensor and remove the with.